macros/src/main/scala/com/gu/contentapi/circe/CirceScroogeMacros.scala

```diff
@@ -135,9 +138,100 @@ object CirceScroogeMacros {
     """
   }
 
-  implicit def encodeThriftStruct[A <: ThriftStruct]: Encoder[A] = macro encodeThriftStruct_impl[A]
+  /**
+    * Macro to produce Decoders for ThriftUnion types.
+    *
+    * A ThriftUnion, U, has a companion object which defines a case class extending U for each member of the union.
+    * Each case class takes a single parameter, whose type is an alias for the actual type contained by that member.
+    *
+    * E.g. for the following thrift definition:
+    * {{{
+    *   union U {
+    *     1: FooStruct foo
+    *     2: BarStruct bar
+    *   }
+    *
+    *   struct T {
+    *     1: U union
+    *   }
+    * }}}
+    *
+    * We may expect the following JSON:
+    * {{{
+    *   {
+    *     union: {
+    *       foo: {
+    *         ...
+    *       }
+    *     }
+    *   }
+    * }}}
+    *
+    * In the scrooge-generated code, for the member foo of union U, there exists a case class:
+    *   {{{
+    *   case class Foo(foo: FooAlias) extends U
+    *   }}}
+    * where FooAlias aliases FooStruct.
+    * We need to match on the single JSON field and determine which member of the union is present.
+    * So the decoder will contain a case statement for the member foo as follows:
+    *   {{{
+    *   case "foo" => c.cursor.downField("foo").flatMap(_.as[FooStruct](decoderForFooStruct).toOption).map(Foo)
+    *   }}}
+    *
+    */
+  implicit def decodeThriftUnion[A <: ThriftUnion]: Decoder[A] = macro decodeThriftUnion_impl[A]
+
+  def decodeThriftUnion_impl[A: c.WeakTypeTag](c: blackbox.Context): c.Tree = {
+    import c.universe._
+
+    val A = weakTypeOf[A]
+
+    val memberClasses: Iterable[Symbol] = A.companion.members.filter { member =>
```

Comment:
For the record `knownDirectSubclasses` [is fixed in scala `2.12.1`](https://github.com/scala/scala/pull/5345) and the fix [will be released in scala in `2.11.9`](https://github.com/scala/scala/pull/5345)